### PR TITLE
Upgrade QPL to v1.1.0 + get maximum jobs from libaccel_config 

### LINKED ISSUE #47877: Errors on building QPL Library
> Make sure that `git diff` result is empty and you've just pulled fresh master. Try cleaning up cmake cache. Just in case, official build instructions are published here: https://clickhouse.com/docs/en/development/build/

**Operating system**

> OS kind or distribution, specific version/release, non-standard kernel if any. If you are trying to build inside virtual machine, please mention it too.
openSUSE 15.4
**Cmake version**
cmake version 3.25.2
**Ninja version**
1.11.1
**Compiler name and version**
clang version 15.0.7
Target: x86_64-suse-linux
Thread model: posix
InstalledDir: /usr/bin
**Full cmake and/or ninja output**
cmake config:
``` bash
cmake -D ENABLE_QPL=ON -D ENABLE_AVX2=ON -D ENABLE_AVX512=ON -D PARALLEL_COMPILE_JOBS=192 -D PARALLEL_LINK_JOBS=192 -D WERROR=OFF -D ENABLE_AVX=ON -D ENABLE_BMI=ON -D ENABLE_BMI2=ON -D ENABLE_AVX2_FOR_SPEC_OP=ON -D ENABLE_AVX512_FOR_SPEC_OP=ON -D ENABLE_SENTRY=OFF -D ENABLE_CLICKHOUSE_TEST=ON ..
```
Errors occured on building QPL library:
```
[116/3977] Building CXX object contrib/qpl-cmake/CMakeFiles/middle_layer_lib.dir/__/qpl/sources/middle-layer/util/awaiter.cpp.o
FAILED: contrib/qpl-cmake/CMakeFiles/middle_layer_lib.dir/__/qpl/sources/middle-layer/util/awaiter.cpp.o 
/usr/bin/ccache /usr/bin/clang++ --target=x86_64-linux-gnu --sysroot=/home/nomad/ClickHouse/cmake/linux/../../contrib/sysroot/linux-x86_64/x86_64-linux-gnu/libc -DQPL_BADARG_CHECK -DQPL_LIB -DQPL_VERSION=\"1.0.0\" -DSTD_EXCEPTION_HAS_STACK_TRACE=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -I/home/nomad/ClickHouse/contrib/qpl-cmake -I/home/nomad/ClickHouse/contrib/qpl/sources/middle-layer -I/home/nomad/ClickHouse/contrib/qpl/include -I/home/nomad/ClickHouse/base/glibc-compatibility/memcpy -I/home/nomad/ClickHouse/contrib/qpl/sources/core-sw/include -I/home/nomad/ClickHouse/contrib/qpl/sources/core-sw/src/include -I/home/nomad/ClickHouse/contrib/qpl/sources/core-sw/src/compression/include -I/home/nomad/ClickHouse/contrib/qpl/sources/isal/include -I/home/nomad/ClickHouse/contrib/qpl/sources/isal/igzip -I/home/nomad/ClickHouse/contrib/qpl/sources/core-iaa/include -isystem /home/nomad/ClickHouse/contrib/llvm-project/libcxx/include -isystem /home/nomad/ClickHouse/contrib/llvm-project/libcxxabi/include -isystem /home/nomad/ClickHouse/contrib/libunwind/include --gcc-toolchain=/home/nomad/ClickHouse/cmake/linux/../../contrib/sysroot/linux-x86_64 -std=c++20 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -fsized-deallocation  -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -mavx -mavx2 -mavx512f -mavx512bw -mavx512vl -mbmi -mbmi2 -fasynchronous-unwind-tables -ffile-prefix-map=/home/nomad/ClickHouse=. -falign-functions=32 -mbranches-within-32B-boundaries -fdiagnostics-absolute-paths -fstrict-vtable-pointers -fexperimental-new-pass-manager -w -O2 -g -DNDEBUG -O3 -g -gdwarf-4  -fno-pie   -D OS_LINUX -nostdinc++ -std=c++2b -MD -MT contrib/qpl-cmake/CMakeFiles/middle_layer_lib.dir/__/qpl/sources/middle-layer/util/awaiter.cpp.o -MF contrib/qpl-cmake/CMakeFiles/middle_layer_lib.dir/__/qpl/sources/middle-layer/util/awaiter.cpp.o.d -o contrib/qpl-cmake/CMakeFiles/middle_layer_lib.dir/__/qpl/sources/middle-layer/util/awaiter.cpp.o -c /home/nomad/ClickHouse/contrib/qpl/sources/middle-layer/util/awaiter.cpp
In file included from /home/nomad/ClickHouse/contrib/qpl/sources/middle-layer/util/awaiter.cpp:14:
/usr/lib64/clang/15.0.7/include/intrin.h:12:15: fatal error: 'intrin.h' file not found
#include_next <intrin.h>
              ^~~~~~~~~~
1 error generated.
```

### PR DESCRIPTION
**Pull request by jinjunzh**
Fixes #47877

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
1. Bump Intel QPL from v1.0.0 to v1.1.0 (fixes build issue #47877) 2. the DEFLATE_QPL codec now respects the maximum hardware jobs returned by libaccel_config

### CODE REVIEW

File: src/Compression/CompressionCodecDeflateQpl.h [hunk 1]
```diff
@@ -24,22 +24,23 @@ class DeflateQplJobHWPool
...
-    static bool tryLockJob(UInt32 index);
-    static void unLockJob(UInt32 index);
+    bool tryLockJob(UInt32 index);
+    void unLockJob(UInt32 index);
 
+    /// size of each job objects
+    UInt32 per_job_size;
     /// Maximum jobs running in parallel supported by IAA hardware
-    static constexpr auto MAX_HW_JOB_NUMBER = 1024;
+    UInt32 hw_jobs_max_number;
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecDeflateQpl.h:37]: Minor/cosmetic: `max` prefix for maximums is fairly common. What about just `max_hw_jobs`?

File: src/Compression/CompressionCodecDeflateQpl.cpp [hunk 2]
```diff
@@ -16,59 +17,72 @@ namespace ErrorCodes
...
-    , distribution(0, MAX_HW_JOB_NUMBER - 1)
+    : hw_jobs_max_number(0)
+    , random_engine(std::random_device()())
 {
     Poco::Logger * log = &Poco::Logger::get("DeflateQplJobHWPool");
-    UInt32 job_size = 0;
     const char * qpl_version = qpl_get_library_version();
 
+    // loop all configured workqueue size to get maximum job number.
+    accfg_ctx *ctx_ptr = nullptr;
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecDeflateQpl.cpp:34]: (cosmetic: \+ space after `*`, also in l. 39 and 40)

File: src/Compression/CompressionCodecDeflateQpl.cpp [hunk 3]
```diff
@@ -16,59 +17,72 @@ namespace ErrorCodes
...
+        while (nullptr != dev_ptr) { // loop all devices
+            for (auto *wq_ptr = accfg_wq_get_first(dev_ptr); \
+                wq_ptr != nullptr; \
+                wq_ptr = accfg_wq_get_next(wq_ptr))
+                hw_jobs_max_number += accfg_wq_get_size(wq_ptr);
+            dev_ptr = accfg_device_get_next(dev_ptr);
+        }
+    }
+
+    if (hw_jobs_max_number == 0)
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecDeflateQpl.cpp:48]: I think a distinction between
- `ctx_status != 0` (l. 36, which seems to indicate a configuration issue or the like) and
- `ctx_status == 0` (l. 36) yet `hw_jobs_max_number == 0` (which seems to indicate correct configuration in general but a lack of hardware resources)

would be helpful. Maybe we can have different logging messages (l. 51) for that?

File: src/Compression/CompressionCodecDeflateQpl.cpp [hunk 4]
```diff
@@ -16,59 +17,72 @@ namespace ErrorCodes
...
 
     job_pool_ready = true;
-    LOG_DEBUG(log, "Hardware-assisted DeflateQpl codec is ready! QPL Version: {}",qpl_version);
+    LOG_DEBUG(log, "Hardware-assisted DeflateQpl codec is ready! QPL Version: {}, hw_jobs_max_number: {}",qpl_version, hw_jobs_max_number);
 }
 
 DeflateQplJobHWPool::~DeflateQplJobHWPool()
 {
-    for (UInt32 i = 0; i < MAX_HW_JOB_NUMBER; ++i)
+    qpl_job * job_ptr = nullptr;
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecDeflateQpl.cpp:79]: Not performance critical, let's move that into the loop (l. 82)

File: src/Compression/CompressionCodecDeflateQpl.cpp [hunk 5]
```diff
@@ -16,59 +17,72 @@ namespace ErrorCodes
...
+    auto ctx_status = accfg_new(&ctx_ptr);
+    if (ctx_status == 0)
+    {
+        auto *dev_ptr = accfg_device_get_first(ctx_ptr);
+        while (nullptr != dev_ptr) { // loop all devices
+            for (auto *wq_ptr = accfg_wq_get_first(dev_ptr); \
+                wq_ptr != nullptr; \
+                wq_ptr = accfg_wq_get_next(wq_ptr))
+                hw_jobs_max_number += accfg_wq_get_size(wq_ptr);
+            dev_ptr = accfg_device_get_next(dev_ptr);
```
[the remark this answers is not in the document]
Comment (jinjunzh) [reply src/Compression/CompressionCodecDeflateQpl.cpp:44]: Either of cases as you mentioned would happen. HW jobs limit depends on both of CPU and user configuration.

File: src/Compression/CompressionCodecDeflateQpl.h (hunk 1 above)
Comment (jinjunzh) [reply src/Compression/CompressionCodecDeflateQpl.h]: Done

File: src/Compression/CompressionCodecDeflateQpl.cpp (hunk 2 above)
Comment (jinjunzh) [reply src/Compression/CompressionCodecDeflateQpl.cpp]: Done

File: src/Compression/CompressionCodecDeflateQpl.cpp [hunk 6]
```diff
@@ -16,59 +17,72 @@ namespace ErrorCodes
...
-    UInt32 job_size = 0;
     const char * qpl_version = qpl_get_library_version();
 
+    // loop all configured workqueue size to get maximum job number.
+    accfg_ctx *ctx_ptr = nullptr;
+    auto ctx_status = accfg_new(&ctx_ptr);
+    if (ctx_status == 0)
+    {
+        auto *dev_ptr = accfg_device_get_first(ctx_ptr);
+        while (nullptr != dev_ptr) { // loop all devices
```
[the remark this answers is not in the document]
Comment (jinjunzh) [reply src/Compression/CompressionCodecDeflateQpl.cpp]: Done

File: src/Compression/CompressionCodecDeflateQpl.cpp [hunk 7]
```diff
@@ -16,59 +17,72 @@ namespace ErrorCodes
...
+            for (auto *wq_ptr = accfg_wq_get_first(dev_ptr); \
```
[the remark this answers is not in the document]
Comment (jinjunzh) [reply src/Compression/CompressionCodecDeflateQpl.cpp]: Done

File: src/Compression/CompressionCodecDeflateQpl.cpp (hunk 3 above)
Comment (jinjunzh) [reply src/Compression/CompressionCodecDeflateQpl.cpp]: Done

File: src/Compression/CompressionCodecDeflateQpl.cpp [hunk 8]
```diff
@@ -16,59 +17,72 @@ namespace ErrorCodes
...
     job_pool_ready = true;
-    LOG_DEBUG(log, "Hardware-assisted DeflateQpl codec is ready! QPL Version: {}",qpl_version);
+    LOG_DEBUG(log, "Hardware-assisted DeflateQpl codec is ready! QPL Version: {}, hw_jobs_max_number: {}",qpl_version, hw_jobs_max_number);
 }
 
 DeflateQplJobHWPool::~DeflateQplJobHWPool()
 {
-    for (UInt32 i = 0; i < MAX_HW_JOB_NUMBER; ++i)
+    qpl_job * job_ptr = nullptr;
+    for (UInt32 index = 0; index < hw_jobs_max_number; ++index)
```
[the remark this answers is not in the document]
Comment (jinjunzh) [reply src/Compression/CompressionCodecDeflateQpl.cpp]: Done

File: src/Compression/CompressionCodecDeflateQpl.cpp (hunk 4 above)
Comment (jinjunzh) [reply src/Compression/CompressionCodecDeflateQpl.cpp]: Done

File: src/Compression/CompressionCodecDeflateQpl.cpp [hunk 9]
```diff
@@ -16,59 +17,72 @@ namespace ErrorCodes
...
+                wq_ptr = accfg_wq_get_next(wq_ptr))
+                hw_jobs_max_number += accfg_wq_get_size(wq_ptr);
+            dev_ptr = accfg_device_get_next(dev_ptr);
+        }
+    }
+
+    if (hw_jobs_max_number == 0)
+    {
+        job_pool_ready = false;
+        LOG_WARNING(log, "Initialization of hardware-assisted DeflateQpl codec failed, falling back to software DeflateQpl codec. Please check if Intel In-Memory Analytics Accelerator (IAA) is properly set up. accfg_context_status: {} ,total_wq_size: {} , QPL Version: {}.", ctx_status, hw_jobs_max_number, qpl_version);
```
[the remark this answers is not in the document]
Comment (jinjunzh) [reply src/Compression/CompressionCodecDeflateQpl.cpp]: Done

**Review by rschu1ze — APPROVED** [review]

**Comment by Tiaonmmn** [comment]
When use cmake setting:
`cmake -D ENABLE_QPL=ON -D ENABLE_AVX2=ON -D ENABLE_AVX512=ON -D PARALLEL_COMPILE_JOBS=128 -D ENABLE_TESTS=ON -D PARALLEL_LINK_JOBS=64 ../
`
Errors occured:
```
[11659/11990] Linking CXX executable utils/check-marks/check-marks
FAILED: utils/check-marks/check-marks
: && /usr/bin/clang++ --target=x86_64-linux-gnu --sysroot=/home/nomad/ClickHouse/cmake/linux/../../contrib/sysroot/linux-x86_64/x86_64-linux-gnu/libc --gcc-toolchain=/home/nomad/ClickHouse/cmake/linux/../../contrib/sysroot/linux-x86_64 -std=c++20 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -Wno-enum-constexpr-conversion -fsized-deallocation  -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -mavx2 -mavx512f -mavx512bw -mavx512vl -fasynchronous-unwind-tables -ffile-prefix-map=/home/nomad/ClickHouse=. -falign-functions=32 -mbranches-within-32B-boundaries -fdiagnostics-absolute-paths -fstrict-vtable-pointers -Wall -Wextra -Wframe-larger-than=65536 -Weverything -Wpedantic -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c++20-compat -Wno-sign-conversion -Wno-implicit-int-conversion -Wno-implicit-int-float-conversion -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -Wno-enum-constexpr-conversion -Wno-unsafe-buffer-usage -O2 -g -DNDEBUG -O3 -g -gdwarf-4  -fno-pie --gcc-toolchain=/home/nomad/ClickHouse/cmake/linux/../../contrib/sysroot/linux-x86_64 --ld-path=/home/nomad/ck-build/ld.lld -rdynamic -Wl,--gdb-index -Wl,--build-id=sha1 -no-pie -Wl,-no-pie    -Xlinker --no-undefined src/CMakeFiles/clickhouse_malloc.dir/Common/malloc.cpp.o utils/check-marks/CMakeFiles/check-marks.dir/main.cpp.o -o utils/check-marks/check-marks  src/libclickhouse_new_delete.a  src/libdbms.a  contrib/boost-cmake/lib_boost_program_options.a  contrib/llvm-project/llvm/lib/libLLVMExecutionEngine.a  contrib/llvm-project/llvm/lib/libLLVMOrcTargetProcess.a  contrib/llvm-project/llvm/lib/libLLVMOrcShared.a  contrib/llvm-project/llvm/lib/libLLVMRuntimeDyld.a  contrib/llvm-project/llvm/lib/libLLVMPasses.a  contrib/llvm-project/llvm/lib/libLLVMCoroutines.a  contrib/llvm-project/llvm/lib/libLLVMObjCARCOpts.a  contrib/llvm-project/llvm/lib/libLLVMipo.a  contrib/llvm-project/llvm/lib/libLLVMFrontendOpenMP.a  contrib/llvm-project/llvm/lib/libLLVMIRReader.a  contrib/llvm-project/llvm/lib/libLLVMAsmParser.a  contrib/llvm-project/llvm/lib/libLLVMLinker.a  contrib/llvm-project/llvm/lib/libLLVMVectorize.a  contrib/llvm-project/llvm/lib/libLLVMX86CodeGen.a  contrib/llvm-project/llvm/lib/libLLVMAsmPrinter.a  contrib/llvm-project/llvm/lib/libLLVMGlobalISel.a  contrib/llvm-project/llvm/lib/libLLVMSelectionDAG.a  contrib/llvm-project/llvm/lib/libLLVMCodeGen.a  contrib/llvm-project/llvm/lib/libLLVMBitWriter.a  contrib/llvm-project/llvm/lib/libLLVMScalarOpts.a  contrib/llvm-project/llvm/lib/libLLVMAggressiveInstCombine.a  contrib/llvm-project/llvm/lib/libLLVMInstCombine.a  contrib/llvm-project/llvm/lib/libLLVMInstrumentation.a  contrib/llvm-project/llvm/lib/libLLVMTransformUtils.a  contrib/llvm-project/llvm/lib/libLLVMTarget.a  contrib/llvm-project/llvm/lib/libLLVMAnalysis.a  contrib/llvm-project/llvm/lib/libLLVMProfileData.a  contrib/llvm-project/llvm/lib/libLLVMSymbolize.a  contrib/llvm-project/llvm/lib/libLLVMDebugInfoDWARF.a  contrib/llvm-project/llvm/lib/libLLVMDebugInfoPDB.a  contrib/llvm-project/llvm/lib/libLLVMObject.a  contrib/llvm-project/llvm/lib/libLLVMBitReader.a  contrib/llvm-project/llvm/lib/libLLVMMCParser.a  contrib/llvm-project/llvm/lib/libLLVMTextAPI.a  contrib/llvm-project/llvm/lib/libLLVMDebugInfoMSF.a  contrib/llvm-project/llvm/lib/libLLVMX86Desc.a  contrib/llvm-project/llvm/lib/libLLVMMCDisassembler.a  contrib/llvm-project/llvm/lib/libLLVMX86Info.a  contrib/llvm-project/llvm/lib/libLLVMMC.a  contrib/llvm-project/llvm/lib/libLLVMDebugInfoCodeView.a  contrib/llvm-project/llvm/lib/libLLVMCFGuard.a  contrib/llvm-project/llvm/lib/libLLVMCore.a  contrib/llvm-project/llvm/lib/libLLVMRemarks.a  contrib/llvm-project/llvm/lib/libLLVMBitstreamReader.a  contrib/llvm-project/llvm/lib/libLLVMBinaryFormat.a  contrib/llvm-project/llvm/lib/libLLVMSupport.a  contrib/llvm-project/llvm/lib/libLLVMDemangle.a  contrib/croaring-cmake/lib_roaring.a  contrib/cppkafka-cmake/lib_cppkafka.a  contrib/librdkafka-cmake/lib_rdkafka.a  contrib/nats-io-cmake/lib_nats_io.a  contrib/cyrus-sasl-cmake/lib_sasl2.a  contrib/nuraft-cmake/lib_nuraft.a  contrib/boost-cmake/lib_boost_coroutine.a  src/Common/Config/libclickhouse_common_config.a  contrib/yaml-cpp-cmake/lib_yaml_cpp.a  src/Common/ZooKeeper/libclickhouse_common_zookeeper.a  src/Dictionaries/Embedded/libclickhouse_dictionaries_embedded.a  src/Parsers/libclickhouse_parsers.a  src/Access/Common/libclickhouse_common_access.a  base/poco/MongoDB/lib_poco_mongodb.a  src/Common/mysqlxx/libmysqlxx.a  src/libclickhouse_common_io.a  contrib/boost-cmake/lib_boost_program_options.a  contrib/jemalloc-cmake/lib_jemalloc.a  contrib/gwpasan-cmake/lib_gwp_asan.a  src/Common/StringUtils/libstring_utils.a  base/widechar_width/libwidechar_width.a  base/base/libcommon.a  contrib/boost-cmake/lib_boost_system.a  base/poco/NetSSL_OpenSSL/lib_poco_net_ssl.a  base/poco/Net/lib_poco_net.a  base/poco/Crypto/lib_poco_crypto.a  base/poco/Util/lib_poco_util.a  base/poco/JSON/lib_poco_json.a  base/poco/JSON/lib_poco_json_pdjson.a  base/poco/XML/lib_poco_xml.a  base/poco/XML/lib_poco_xml_expat.a  contrib/replxx-cmake/lib_replxx.a  contrib/cctz-cmake/lib_cctz.a  -Wl,--whole-archive /home/nomad/ck-build/contrib/cctz-cmake/libtzdata.a -Wl,--no-whole-archive  contrib/fmtlib-cmake/lib_fmt.a  contrib/dragonbox-cmake/lib_dragonbox_to_chars.a  contrib/cityhash102/lib_cityhash.a  contrib/re2-cmake/libre2_st.a  base/poco/Foundation/lib_poco_foundation.a  base/poco/Foundation/lib_poco_foundation_pcre.a  contrib/libcpuid-cmake/lib_cpuid.a  contrib/aws-cmake/lib_aws.a  contrib/liburing-cmake/lib_liburing.a  contrib/minizip-ng-cmake/lib_minizip.a  contrib/xz-cmake/lib_liblzma.a  contrib/bzip2-cmake/lib_bzip2.a  contrib/mariadb-connector-c-cmake/lib_mariadbclient.a  contrib/icu-cmake/lib_icui18n.a  contrib/icu-cmake/lib_icuuc.a  contrib/icu-cmake/lib_icudata.a  contrib/capnproto-cmake/lib_capnpc.a  contrib/capnproto-cmake/lib_capnp.a  contrib/capnproto-cmake/lib_kj.a  contrib/arrow-cmake/lib_parquet.a  contrib/arrow-cmake/lib_arrow.a  contrib/boost-cmake/lib_boost_filesystem.a  contrib/double-conversion-cmake/lib_double-conversion.a  contrib/brotli-cmake/lib_brotli.a  contrib/flatbuffers/libflatbuffers.a  contrib/arrow-cmake/lib_orc.a  contrib/avro-cmake/lib_avrocpp.a  contrib/boost-cmake/lib_boost_iostreams.a  contrib/openldap-cmake/lib_ldap_r.a  contrib/openldap-cmake/lib_lber.a  src/Server/grpc_protos/libclickhouse_grpc_protos.a  contrib/grpc/libgrpc++.a  contrib/grpc/libgrpc.a  contrib/re2-cmake/libre2.a  contrib/c-ares-cmake/lib_c-ares.a  contrib/abseil-cpp/absl/status/libabsl_status.a  contrib/grpc/libaddress_sorting.a  contrib/grpc/libupb.a  contrib/grpc/libgpr.a  contrib/libhdfs3-cmake/lib_hdfs3.a  contrib/protobuf-cmake/lib_libprotobuf.a  contrib/libgsasl-cmake/lib_gsasl.a  contrib/krb5-cmake/lib_krb5.a  contrib/isa-l-cmake/lib_isal.a  contrib/hive-metastore-cmake/lib_hivemetastore.a  contrib/thrift-cmake/lib_thrift.a  contrib/azure-cmake/lib_azure_sdk.a  contrib/libxml2-cmake/lib_libxml2.a  contrib/curl-cmake/lib_curl.a  contrib/s2geometry-cmake/lib_s2.a  contrib/abseil-cpp/absl/container/libabsl_raw_hash_set.a  contrib/abseil-cpp/absl/container/libabsl_hashtablez_sampler.a  contrib/abseil-cpp/absl/strings/libabsl_cord.a  contrib/abseil-cpp/absl/strings/libabsl_cordz_info.a  contrib/abseil-cpp/absl/strings/libabsl_cord_internal.a  contrib/abseil-cpp/absl/strings/libabsl_cordz_functions.a  contrib/abseil-cpp/absl/profiling/libabsl_exponential_biased.a  contrib/abseil-cpp/absl/strings/libabsl_cordz_handle.a  contrib/abseil-cpp/absl/synchronization/libabsl_synchronization.a  contrib/abseil-cpp/absl/debugging/libabsl_stacktrace.a  contrib/abseil-cpp/absl/debugging/libabsl_symbolize.a  contrib/abseil-cpp/absl/debugging/libabsl_debugging_internal.a  contrib/abseil-cpp/absl/debugging/libabsl_demangle_internal.a  contrib/abseil-cpp/absl/synchronization/libabsl_graphcycles_internal.a  contrib/abseil-cpp/absl/time/libabsl_time.a  contrib/abseil-cpp/absl/time/libabsl_civil_time.a  contrib/abseil-cpp/absl/time/libabsl_time_zone.a  contrib/abseil-cpp/absl/base/libabsl_malloc_internal.a  contrib/abseil-cpp/absl/hash/libabsl_hash.a  contrib/abseil-cpp/absl/hash/libabsl_city.a  contrib/abseil-cpp/absl/types/libabsl_bad_optional_access.a  contrib/abseil-cpp/absl/types/libabsl_bad_variant_access.a  contrib/abseil-cpp/absl/hash/libabsl_low_level_hash.a  contrib/abseil-cpp/absl/strings/libabsl_str_format_internal.a  contrib/abseil-cpp/absl/strings/libabsl_strings.a  contrib/abseil-cpp/absl/numeric/libabsl_int128.a  contrib/abseil-cpp/absl/base/libabsl_throw_delegate.a  contrib/abseil-cpp/absl/strings/libabsl_strings_internal.a  contrib/abseil-cpp/absl/base/libabsl_base.a  contrib/abseil-cpp/absl/base/libabsl_spinlock_wait.a  -lrt  contrib/abseil-cpp/absl/base/libabsl_raw_logging_internal.a  contrib/abseil-cpp/absl/base/libabsl_log_severity.a  contrib/amqpcpp-cmake/lib_amqp-cpp.a  contrib/libuv-cmake/lib_uv.a  -lrt  contrib/sqlite-cmake/lib_sqlite.a  contrib/rocksdb-cmake/lib_rocksdb.a  contrib/lz4-cmake/lib_lz4.a  contrib/zstd-cmake/lib_zstd.a  contrib/zlib-ng-cmake/lib_zlib.a  contrib/snappy-cmake/lib_snappy.a  contrib/libpqxx-cmake/lib_libpqxx.a  contrib/libpq-cmake/lib_libpq.a  contrib/boringssl-cmake/lib_ssl.a  contrib/boringssl-cmake/lib_crypto.a  -lpthread  contrib/qpl-cmake/lib_qpl.a  contrib/qpl-cmake/libaccel-config.a  -ldl  contrib/boost-cmake/lib_boost_context.a  contrib/libstemmer-c-cmake/lib_stemmer.a  contrib/wordnet-blast-cmake/lib_wnb.a  contrib/boost-cmake/lib_boost_graph.a  contrib/boost-cmake/lib_boost_regex.a  contrib/lemmagen-c-cmake/lib_lemmagen.a  -Wl,--whole-archive /home/nomad/ck-build/contrib/nlp-data-cmake/libnlp_dictionaries.a -Wl,--no-whole-archive  contrib/ulid-c-cmake/lib_ulid.a  contrib/simdjson-cmake/lib_simdjson.a  contrib/consistent-hashing/lib_consistent_hashing.a  rust/skim/RelWithDebInfo/lib_ch_rust_skim_rust.a  rust/skim/lib_ch_rust_skim_ffi.a  base/glibc-compatibility/libglibc-compatibility.a  base/glibc-compatibility/memcpy/libmemcpy.a  base/glibc-compatibility/libglibc-compatibility.a  base/glibc-compatibility/memcpy/libmemcpy.a  -Wl,--start-group  contrib/libcxx-cmake/libcxx.a  contrib/libcxxabi-cmake/libcxxabi.a  contrib/libunwind-cmake/libunwind.a  -Wl,--end-group  -nodefaultlibs /usr/lib64/clang/16/lib/linux/libclang_rt.builtins-x86_64.a  -lc -lm -lrt -lpthread -ldl && :
ld.lld: error: duplicate symbol: isal_update_histogram_base
>>> defined at huff_codes.c:661 (./contrib/isa-l/igzip/huff_codes.c:661)
>>>            huff_codes.c.o:(isal_update_histogram_base) in archive contrib/isa-l-cmake/lib_isal.a
>>> defined at huff_codes.c:654 (./contrib/qpl/sources/isal/igzip/huff_codes.c:654)
>>>            huff_codes.c.o:(.text+0x0) in archive contrib/qpl-cmake/lib_qpl.a

ld.lld: error: duplicate symbol: isal_create_hufftables
>>> defined at huff_codes.c:1377 (./contrib/isa-l/igzip/huff_codes.c:1377)
>>>            huff_codes.c.o:(isal_create_hufftables) in archive contrib/isa-l-cmake/lib_isal.a
>>> defined at huff_codes.c:1371 (./contrib/qpl/sources/isal/igzip/huff_codes.c:1371)
>>>            huff_codes.c.o:(.text+0x3E0) in archive contrib/qpl-cmake/lib_qpl.a

ld.lld: error: duplicate symbol: isal_create_hufftables_subset
>>> defined at huff_codes.c:1467 (./contrib/isa-l/igzip/huff_codes.c:1467)
>>>            huff_codes.c.o:(isal_create_hufftables_subset) in archive contrib/isa-l-cmake/lib_isal.a
>>> defined at huff_codes.c:1474 (./contrib/qpl/sources/isal/igzip/huff_codes.c:1474)
>>>            huff_codes.c.o:(.text+0x31C0) in archive contrib/qpl-cmake/lib_qpl.a

ld.lld: error: duplicate symbol: create_hufftables_icf
>>> defined at huff_codes.c:1591 (./contrib/isa-l/igzip/huff_codes.c:1591)
>>>            huff_codes.c.o:(create_hufftables_icf) in archive contrib/isa-l-cmake/lib_isal.a
>>> defined at huff_codes.c:1624 (./contrib/qpl/sources/isal/igzip/huff_codes.c:1624)
>>>            huff_codes.c.o:(.text+0x4F60) in archive contrib/qpl-cmake/lib_qpl.a
clang-16.0: error: linker command failed with exit code 1 (use -v to see invocation)
```
But,if use cmake setting:
`cmake -D ENABLE_QPL=ON -D ENABLE_AVX2=ON -D ENABLE_AVX512=ON -D PARALLEL_COMPILE_JOBS=128 -D ENABLE_TESTS=OFF -D PARALLEL_LINK_JOBS=64 ../ClickHouse/
`
Build succeed.
I think it's related to #48833 .

**Comment by jinjunzh** [comment]
@rschu1ze  what's your suggestion for this conflict issue between isa-l in ./contrib/QPL and ./contrib/isa-l ?

**Comment by rschu1ze** [comment]
The cause of the problem is that QPL comes with its [own dump](https://github.com/intel/qpl/tree/develop/sources/isal) of isa-l (it would be interesting to understand why - just convenience or does it deviate in some ways from upstream isa-l?). ClickHouse now unconditionally links upstream isa-l as per #48833 which leans to duplicate symbols when the QPL codec is enabled.

There is precedence for that (e.g. #43259: cJSON linked by two unrelated submodules), and besides hacks (configuring the library to not export symbols, using PRIVATE linkage in CMake, ...) the only *real* fix is to build/link isa-l only once. So the best direction IMHO is to make upstream QPL not include a copy of isa-l but have isa-l as a prerequisite. "qpl-cmake/CMakeLists.txt" could then simply do `target_link_libraries(_qpl ch_contrib::isal)` (see contrib/isa-l-cmake/CMakeLists.txt) to satisfy the isa-l requirements. Note that the issue is not specific to ClickHouse, every program linking QPL and isa-l will run into this problem.

**Comment by rschu1ze** [comment]
Let's continue discussion on the new issue in #49296. It actually existed before this PR, the real cause was #48833.

**Pull request merged**